man/dmlogger.1
.TH "DMLOGGER" "1" "" "Version 2.0.0" "User Commands"
.SH NAME
dmlogger \(en inserts collected log messages into database
.SH SYNOPSIS
\f[B]dmlogger\f[R] \-\-help
.PP
\f[B]dmlogger\f[R] \-\-version
.PP
\f[B]dmlogger\f[R] \-\-\f[B]node\f[R] \f[I]id\f[R]
[\-\-\f[B]database\f[R] \f[I]file\f[R]] [\-\-\f[B]output\f[R]
\f[I]file\f[R]] [\-\-\f[B]minlevel\f[R] \f[I]n\f[R]] [\-\-\f[B]ipc\f[R]]
[\-\-\f[B]verbose\f[R]]
.PP
\f[B]dmlogger\f[R] \-\-\f[B]config\f[R] \f[I]file\f[R]
[\-\-\f[B]name\f[R] \f[I]name\f[R]]
.SH DESCRIPTION
The \f[B]dmlogger\f[R] program collects log messages from a POSIX
message queue and inserts them into a SQLite database.
The name of the message queue matches the given \f[B]dmlogger\f[R] name,
by default \f[CR]/dmlogger\f[R].
.PP
If a minimum log level is selected, only logs of a level greater equal
the minimum are written to log file or log database.
Log messages with lower level are discarded.
.PP
The option \f[CR]\-\-ipc\f[R] enables process synchronisation via POSIX
semaphores.
The semaphore value is incremented from 0 to 1 whenever a new log has
been delivered.
The name of the semaphore equals the \f[B]dmlogger\f[R] name with
leading \f[CR]/\f[R].
Only a single process shall wait for the semaphore unless round\-robin
IPC is desired.
This option may be used to synchronise incoming log messages
automatically with a remote RPC API server.
For example, pass the \f[B]dmlogger\f[R] name to \f[I]dmsync(1)\f[R]
through command\-line argument \f[CR]\-\-wait\f[R].
The semaphore will then trigger the log synchronisation.
.SH OPTIONS
.TP
\f[B]\-\-config\f[R], \f[B]\-c\f[R] \f[I]file\f[R]
Path to configuration file.
.TP
\f[B]\-\-database\f[R], \f[B]\-d\f[R] \f[I]file\f[R]
Path to SQLite log database.
.TP
\f[B]\-\-help\f[R], \f[B]\-h\f[R]
Print available command\-line arguments and quit.
.TP
\f[B]\-\-ipc\f[R], \f[B]\-Q\f[R]
Use POSIX semaphore for process synchronisation.
The name of the semaphore matches the instance name (with leading
slash).
The semaphore is set to 1 each time a log message has been received.
Only a single process shall wait for this semaphore.
.TP
\f[B]\-\-minlevel\f[R], \f[B]\-L\f[R] \f[I]level\f[R]
Minimum level for a log to be stored in the database, from 1
(\f[CR]LL_DEBUG\f[R]) to 5 (\f[CR]LL_CRITICAL\f[R]).
The argument may be an integer or name string.
.TP
\f[B]\-\-name\f[R], \f[B]\-n\f[R] \f[I]name\f[R]
Name of logger instance, configuration, POSIX message queue, and POSIX
semaphore (default is \f[CR]dmlogger\f[R]).
.TP
\f[B]\-\-node\f[R], \f[B]\-N\f[R] \f[I]id\f[R]
Node id.
.TP
\f[B]\-\-output\f[R], \f[B]\-o\f[R] \f[I]file\f[R]
Path of log file.
If the file does not exist, a new log file will be created.
Logs are appended to the end of the file.
If set to \f[CR]\-\f[R], logs are printed to \f[I]stderr\f[R] (equals
argument \f[CR]\-\-verbose\f[R]).
.TP
\f[B]\-\-verbose\f[R], \f[B]\-V\f[R]
Print received logs to \f[I]stderr\f[R].
.TP
\f[B]\-\-version\f[R], \f[B]\-v\f[R]
Print version information and quit.
.SH EXIT STATUS
.TP
\f[B]0\f[R]
Success.
Program executed without errors.
.TP
\f[B]1\f[R]
Failure.
Program execution failed.
.SH ENVIRONMENT
.TP
\f[B]NO_COLOR\f[R]
Disable ANSI colour output.
.SH EXAMPLE
Create a POSIX message queue \f[CR]/dmlogger\f[R] and wait for incoming
logs:
.IP
.EX
$ dmlogger \-N dummy\-node \-d /var/dmpack/log.db \-L info
.EE
.PP
Received logs are inserted only if they are of level \f[I]info\f[R] or
higher.
.SH SEE ALSO
\f[I]dmlog(1)\f[R], \f[I]dmsync(1)\f[R]


CHANGELOG.md
# Change Log

All notable changes to the project will be documented in this file.

## [2.0.0] – Unreleased

## Library

* `Added` error codes
* `Added` _fortran-nng_ interface bindings to NNG to `vendor/`
* `Added` module `dm_arg_parser` (outsourced from `dm_arg`)
* `Added` module `dm_group` for observation groups
* `Added` module `dm_ipc` for socket access
* `Added` module `dm_ipc_async` for asynchronous I/O
* `Added` module `dm_ipc_disco` for service discovery
* `Added` module `dm_ipc_message` for message handling
* `Added` module `dm_ipc_mutex` for mutex access
* `Added` module `dm_ipc_thread` for thread access
* `Added` module `dm_ipc_type` of IPC enumerators
* `Added` module `dm_job_list` (outsourced from `dm_job`)
* `Changed` names of modules `dm_fifo`, `dm_mqueue`, `dm_mqueue_util`, `dm_pipe`,
  `dm_sem`, `dm_signal`, `dm_thread`, `dm_tty` to include `posix` prefix
* `Changed` name of module `dm_system` to `dm_posix`
* `Changed` database schema due to new data models
* `Changed` default database suffix from `.sqlite` to `.db`
* `Changed` GeoCOM API for Lua to accept prototype observation as first argument
* `Changed` model of observation data structure `dm_observ`
* `Deleted` model of request data strucutre `dm_request`

## Programs

* `Changed` response value units in `dmved`

## Documentation

* Converted user guide from AsciiDoc to Markdown
* Converted man pages from AsciiDoc to Markdown

## [1.0.0] – Unreleased

## Library

* `Added` Bessel, Butterworth, Chebyshev filters to `dm_filter`

## [0.9.9] – 2026-01-11

### Library

* `Added` module `dm_coord` for coordinate transformation
* `Added` module `dm_filter` for filtering
* `Added` module `dm_ghostscript` for Ghostscript access
* `Added` module `dm_netstring` for netstring parsing
* `Added` module `dm_statistics` of statistics functions
* `Added` file tree size function to module `dm_file`
* `Added` file modification date/time to function `dm_file_touch()` in `dm_file`
* `Added` format `FORMAT_TSV` for tab-separated values to `dm_format`
* `Added` swap routines to `dm_util`
* `Changed` error code order for future additions
* `Changed` default database cache size of `dm_db_open()` to improve performance

### Programs

* `Added` author option to `dmreport`
* `Added` meta data to PDF output of `dmreport`
* `Fixed` reading of options in `dmdbctl`
* `Changed` last modified date/time to last log date/time in output of `dmfeed`

## [0.9.8] – 2025-08-30

### Library

* `Added` module `dm_image` for image handling
* `Added` module `dm_js` for JavaScript generation
* `Added` module `dm_roff` for GNU roff abstraction
* `Added` module `dm_serial` for derived type serialisation
* `Added` module `dm_transfer` for HTTP-RPC API file transfers
* `Added` GeoJSON Feature Collection procedures to `dm_geojson`
* `Added` image upload to `dm_rpc`
* `Added` HTTP request and response headers to `dm_rpc`
* `Added` HTTP response headers to `dm_fcgi`
* `Added` terminals `gpic`, `postscript`, and `sixeltek` to `dm_plot`
* `Added` image and transfer functions to `dm_db_api`
* `Changed` structure of database abstraction layer, added modules `dm_db_api`, `dm_db_count`, `dm_db_json`, `dm_db_pragma`, and `dm_db_row`
* `Changed` API of modules `dm_arg` and `dm_config` to object-oriented
* `Changed` query parameter API in `dm_cgi`
* `Changed` Git submodules to copies of third-party dependencies
* `Fixed` reading from pipe in `dm_pipe`
* `Fixed` unsigned type conversion in `dm_c`
* `Fixed` return code handling of database select functions

### Programs

* `Added` program `dmcamera` for capturing of camera images
* `Added` program `dmupload` for image upload to HTTP-RPC API
* `Added` scale factor for response values to `dmreport`
* `Added` PDF and PostScript output to `dmreport`
* `Added` message queue reading to `dmmb`
* `Added` image upload to `dmapi`
* `Added` image viewer to `dmweb`
* `Added` environment variables `DM_IMAGE_DB`, `DM_IMAGE_DIR` to `dmapi` and `dmweb`
* `Added` log file output to `dmlogger`
* `Changed` environment variables in `dmapi` and `dmweb` to `DM_BEAT_DB`, `DM_LOG_DB`, `DM_OBSERV_DB`
* `Changed` derived type serialisation in `dmapi` and `dmexport`
* `Changed` database access in `dmexport`
* `Changed` configuration of `dmgrc`

## [0.9.7] – 2025-04-05

### Library

* `Added` module `dm_c` for C interoperability
* `Added` module `dm_camera` for webcam and RTSP access (FFmpeg)
* `Added` module `dm_crypto` of cryptographic hash functions
* `Added` module `dm_db_query` for SQL query building
* `Added` module `dm_dwd` for DWD API format parsing
* `Added` module `dm_dwd_api` for DWD API access (libcurl)
* `Added` module `dm_ftp` for FTP transfer (libcurl)
* `Added` module `dm_freebsd` for system status access on FreeBSD
* `Added` module `dm_geojson` for GeoJSON serialisation
* `Added` module `dm_gm` for GraphicsMagick access
* `Added` module `dm_im` for XMPP connectivity (libstrophe)
* `Added` module `dm_linux` for system status access on Linux
* `Added` module `dm_modbus` for Modbus RTU/TCP connectivity (libmodbus)
* `Added` module `dm_ve` for Victron Energy VE.Direct protocol support
* `Added` library build date and system type parameters (CPP, FPP)
* `Added` Linux and FreeBSD abstractions in module `dm_system`
* `Changed` refactored database abstraction layer
* `Changed` refactored database schema
* `Changed` refactored command-line argument parsing
* `Changed` refactored logical function names
* `Fixed` wrong baud rate in GeoCOM class

### Programs

* `Added` program `dmbot` for remote control of clients through XMPP
* `Added` program `dmdwd` for Deutscher Wetterdienst (DWD) API monitoring
* `Added` program `dmmb` for sensor monitoring through Modbus RTU/TCP
* `Added` program `dmmbctl` for Modbus RTU/TCP command-line control
* `Added` program `dmsystem` for system monitoring
* `Added` program `dmved` for MPPT and battery monitoring (VE.Direct)
* `Added` interactive map view to `dmweb` (Leaflet)

### Documentation


app/dmlogger.f90
! dmlogger.f90
!
! Author:  Philipp Engel
! Licence: ISC
program dmlogger
    !! The logger program collects log messages from a POSIX message queue and
    !! writes them log file or log database.
    use :: dmpack
    implicit none (type, external)

    character(*), parameter :: APP_NAME  = 'dmlogger'
    integer,      parameter :: APP_MAJOR = 2
    integer,      parameter :: APP_MINOR = 0
    integer,      parameter :: APP_PATCH = 0

    integer, parameter :: APP_DB_NSTEPS  = 500                !! Number of steps before database is optimised.
    integer, parameter :: APP_DB_TIMEOUT = DB_TIMEOUT_DEFAULT !! SQLite 3 busy timeout in mseconds.
    integer, parameter :: APP_UNIT_NONE  = -99999             !! Default file unit.

    type :: app_type
        !! Application settings.
        character(ID_LEN)        :: name      = APP_NAME !! Name of logger instance and POSIX semaphore.
        character(FILE_PATH_LEN) :: config    = ' '      !! Path to configuration file.
        character(FILE_PATH_LEN) :: database  = ' '      !! Path to SQLite database file.
        character(FILE_PATH_LEN) :: output    = ' '      !! Path to output file.
        character(NODE_ID_LEN)   :: node_id   = ' '      !! Node id.
        integer                  :: min_level = LL_INFO  !! Minimum level for a log to be stored in the database.
        logical                  :: ipc       = .false.  !! Use POSIX semaphore for process synchronisation.
        logical                  :: verbose   = .false.  !! Print debug messages to stderr.
    end type app_type

    class(logger_class), pointer :: logger ! Logger object.

    integer                    :: rc     ! Return code.
    integer                    :: unit   ! Log file unit.
    type(app_type)             :: app    ! App configuration.
    type(db_type)              :: db     ! Global database handle.
    type(posix_mqueue_type)    :: mqueue ! Global POSIX message queue handle.
    type(posix_sem_named_type) :: sem    ! Global POSIX semaphore handle.

    ! Disable output to log file.
    unit = APP_UNIT_NONE

    ! Initialise DMPACK.
    call dm_init()

    ! Get command-line arguments, read options from configuration file.
    rc = read_args(app)
    if (dm_is_error(rc)) call dm_stop(STOP_FAILURE)

    ! Initialise logger.
    logger => dm_logger_get_default()
    call logger%configure(name    = app%name,    & ! Name of global logger.
                          node_id = app%node_id, & ! Sensor node id.
                          source  = app%name,    & ! Application name.
                          ipc     = .false.,     & ! Don't send logs via IPC.
                          verbose = app%verbose)   ! Prints logs to terminal.
    call logger%info('started ' // APP_NAME)

    call init(app, db, mqueue, sem, unit, rc)
    if (dm_is_error(rc)) call shutdown(rc)

    call run(app, db, mqueue, sem, unit, rc)
    call shutdown(rc)
contains
    subroutine shutdown(error)
        !! Cleans up and stops program.
        integer, intent(in) :: error !! DMPACK error code.

        integer :: rc, stat

        stat = merge(STOP_FAILURE, STOP_SUCCESS, dm_is_error(error))

        if (dm_db_is_connected(db)) then
            call dm_db_close(db, error=rc)
            if (dm_is_error(rc)) call logger%error('failed to close database', error=rc)
        end if

        if (unit /= APP_UNIT_NONE) close (unit)

        call dm_posix_mqueue_close(mqueue, error=rc)
        if (dm_is_error(rc)) call logger%error('failed to close mqueue /' // app%name, error=rc)

        call dm_posix_mqueue_unlink(mqueue, error=rc)
        if (dm_is_error(rc)) call logger%error('failed to unlink mqueue /' // app%name, error=rc)

        if (app%ipc) then
            call dm_posix_sem_close(sem, error=rc)
            if (dm_is_error(rc)) call logger%error('failed to close semaphore /' // app%name, error=rc)

            call dm_posix_sem_unlink(sem, error=rc)
            if (dm_is_error(rc)) call logger%error('failed to unlink semaphore /' // app%name, error=rc)
        end if

        call logger%info('stopped ' // APP_NAME, error=error)
        call dm_stop(stat)
    end subroutine shutdown

    subroutine init(app, db, mqueue, sem, unit, error)
        !! Opens database, log file, message queue, and semaphore.
        type(app_type),             intent(inout) :: app    !! App settings.
        type(db_type),              intent(inout) :: db     !! Log database.
        type(posix_mqueue_type),    intent(inout) :: mqueue !! Message queue.
        type(posix_sem_named_type), intent(inout) :: sem    !! Semaphore.
        integer,                    intent(inout) :: unit   !! File unit.
        integer,                    intent(out)   :: error  !! Error code.

        init_block: block
            integer :: stat

            ! Open SQLite database.
            if (dm_string_has(app%database)) then
                rc = dm_db_open(db, path=app%database, timeout=APP_DB_TIMEOUT)

                if (dm_is_error(rc)) then
                    call logger%error('failed to open database ' // app%database, error=rc)
                    exit init_block
                end if

                call logger%debug('opened database ' // app%database)

                if (.not. dm_db_table_has_logs(db)) then
                    call logger%error('missing table in database ' // app%database, error=E_INVALID)
                    exit init_block
                end if
            end if

            ! Open log file for writing.
            if (dm_string_has(app%output) .and. app%output /= '-') then
                if (dm_file_is_fifo(app%output)) then
                    ! Open named pipe.
                    open (access='stream', action='write', file=trim(app%output), form='formatted', iostat=stat, newunit=unit, status='old')
                else
                    ! Open regular file.
                    open (action='write', file=trim(app%output), iostat=stat, newunit=unit, position='append', status='unknown')
                end if

                if (stat /= 0) rc = E_IO

                if (dm_is_error(rc)) then
                    call logger%error('failed to open log file ' // app%output, error=rc)
                    return
                end if
            end if

            ! Open log message queue for reading.
            rc = dm_posix_mqueue_open(mqueue, type=TYPE_LOG, name=app%name, access=POSIX_MQUEUE_RDONLY)

            if (dm_is_error(rc)) then
                call logger%error('failed to open mqueue /' // trim(app%name) // ': ' // dm_posix_error_message(), error=rc)
                exit init_block
            end if

            call logger%debug('opened mqueue ' // app%name)

            ! Create semaphore for IPC.
            if (app%ipc) then
                rc = dm_posix_sem_open(sem, app%name, value=0, create=.true.)

                if (dm_is_error(rc)) then
                    call logger%error('failed to open semaphore /' // app%name, error=rc)
                    exit init_block
                end if

                call logger%debug('opened semaphore ' // app%name)
            end if

            ! Register signal handler.
            call dm_posix_signal_register(signal_callback)
        end block init_block

        error = rc
    end subroutine init

    subroutine output(log, unit)
        !! Append log to file `path`.
        type(log_type), intent(inout) :: log  !! Log to write.
        integer,        intent(in)    :: unit !! File unit.

        integer :: rc

        if (unit == APP_UNIT_NONE) return
        call logger%out(log, unit, error=rc)
        if (dm_is_error(rc)) call dm_error_out(rc, 'failed to write to log file')
    end subroutine output

    subroutine run(app, db, mqueue, sem, unit, error)
        !! Writes received logs to log file and/or log database. The given
        !! message queue has to be opened already.
        type(app_type),             intent(inout) :: app    !! App settings.
        type(db_type),              intent(inout) :: db     !! Log database.
        type(posix_mqueue_type),    intent(inout) :: mqueue !! Message queue.
        type(posix_sem_named_type), intent(inout) :: sem    !! Semaphore.
        integer,                    intent(inout) :: unit   !! Log file unit.
        integer,                    intent(out)   :: error  !! Error code.

        integer        :: rc, steps, value
        logical        :: has_db, has_file
        type(log_type) :: log

        has_db   = dm_db_is_connected(db)
        has_file = (unit /= APP_UNIT_NONE .and. app%output /= '-')

        if (has_db) then
            call logger%debug('writing logs of level >= ' // trim(LOG_LEVEL_NAMES(app%min_level)) // ' to database ' // app%database)
        else
            call logger%debug('database storage is disabled')
        end if

        if (has_file) then
            call logger%debug('writing logs of level >= ' // trim(LOG_LEVEL_NAMES(app%min_level)) // ' to file ' // app%output)
        else
            call logger%debug('file storage is disabled')
        end if

        call logger%debug('waiting for log on mqueue /' // trim(app%name))

        steps = 0

        ipc_loop: do
            ! Blocking read from POSIX message queue.
            rc = dm_posix_mqueue_read(mqueue, log)

            if (dm_is_error(rc)) then
                call logger%error('failed to read from mqueue /' // trim(app%name) // ', next attempt in 30 sec', error=rc)
                call dm_sleep(30)
                cycle ipc_loop
            end if

            ! Replace missing or invalid sensor node id.
            if (.not. dm_id_is_valid(log%node_id)) log%node_id = app%node_id

            ! Print log to standard error.
            if (app%verbose) call logger%out(log)

            ! Skip if log level is lower than minimum level required for storage.
            if (log%level < app%min_level) cycle ipc_loop

            ! Write log to file.
            if (has_file) call output(log, unit)

            ! Skip if database storage is disabled.
            if (.not. has_db) cycle ipc_loop

            ! Skip if log already exists.
            if (dm_db_has_log(db, log%id)) then
                call logger%warning('log ' // trim(log%id) // ' exists', error=E_EXIST)
                cycle ipc_loop
            end if

            ! Add log to database.
            db_loop: do
                rc = dm_db_insert(db, log)

                if (dm_is_error(rc)) then
                    ! Get more precise database error.
                    rc = dm_db_error(db)

                    if (rc == E_DB_BUSY) then
                        call logger%warning('database busy', error=rc)
                        call dm_db_sleep(APP_DB_TIMEOUT)
                        cycle db_loop
                    end if

                    call logger%error('failed to insert log ' // trim(log%id) // ': ' // dm_db_error_message(db), error=rc)
                    exit db_loop
                end if

                if (app%ipc) then
                    ! Post semaphore.
                    rc = dm_posix_sem_value(sem, value)
                    if (value == 0) rc = dm_posix_sem_post(sem)
                end if

                exit db_loop
            end do db_loop

            ! Optimise database.
            if (steps == 0) then
                rc = dm_db_optimize(db)

                if (dm_is_error(rc)) then
                    call logger%error('failed to optimize database', error=rc)
                else
                    call logger%debug('optimized database')
                end if
            end if

            ! Increase optimise step counter.
            steps = modulo(dm_inc(steps), APP_DB_NSTEPS)
        end do ipc_loop

        error = rc
    end subroutine run

    ! **************************************************************************
    ! COMMAND-LINE ARGUMENTS AND CONFIGURATION FILE.
    ! **************************************************************************
    integer function read_args(app) result(rc)
        !! Reads command-line arguments and settings from configuration file.
        type(app_type), intent(out) :: app

        type(arg_parser_class) :: parser

        ! Required and optional command-line arguments.
        call parser%add('name',     short='n', type=ARG_TYPE_ID)      ! -n, --name <id>
        call parser%add('config',   short='c', type=ARG_TYPE_FILE)    ! -c, --config <path>
        call parser%add('database', short='d', type=ARG_TYPE_DATABASE, exist=.true.) ! -d, --database <path>
        call parser%add('output',   short='o', type=ARG_TYPE_FILE)    ! -o, --output <path>
        call parser%add('node',     short='N', type=ARG_TYPE_ID)      ! -N, --node <id>
        call parser%add('minlevel', short='L', type=ARG_TYPE_LEVEL)   ! -L, --minlevel <n>
        call parser%add('ipc',      short='Q', type=ARG_TYPE_LOGICAL) ! -Q, --ipc
        call parser%add('verbose',  short='V', type=ARG_TYPE_LOGICAL) ! -V, --verbose

        ! Read all command-line arguments.
        rc = parser%read(version_callback)
        if (dm_is_error(rc)) return

        call parser%get('name',   app%name)
        call parser%get('config', app%config)

        ! Read configuration from file.
        rc = read_config(app)
        if (dm_is_error(rc)) return

        ! Overwrite settings.
        call parser%get('database', app%database)
        call parser%get('output',   app%output)
        call parser%get('node',     app%node_id)
        call parser%get('minlevel', app%min_level)
        call parser%get('ipc',      app%ipc)
        call parser%get('verbose',  app%verbose)

        app%verbose = (app%verbose .or. app%output == '-')
        rc = validate(app)
    end function read_args

    integer function read_config(app) result(rc)
        !! Reads configuration from (Lua) file if path is not emty.
        type(app_type), intent(inout) :: app !! App type.

        type(config_class) :: config

        rc = E_NONE
        if (.not. dm_string_has(app%config)) return

        rc = config%open(app%config, app%name)

        if (dm_is_ok(rc)) then
            call config%get('database', app%database)
            call config%get('output',   app%output)
            call config%get('node',     app%node_id)
            call config%get('minlevel', app%min_level)
            call config%get('ipc',      app%ipc)
            call config%get('verbose',  app%verbose)
        end if

        call config%close()
    end function read_config

    integer function validate(app) result(rc)
        !! Validates options and prints error messages.
        type(app_type), intent(inout) :: app !! App type.

        rc = E_INVALID

        if (.not. dm_id_is_valid(app%name)) then
            call dm_error_out(rc, 'invalid name')
            return
        end if

        if (.not. dm_id_is_valid(app%node_id)) then
            call dm_error_out(rc, 'invalid or missing node id')
            return
        end if

        if (dm_string_has(app%database) .and. .not. dm_file_exists(app%database)) then
            call dm_error_out(rc, 'database does not exist')
            return
        end if

        if (.not. dm_string_has(app%database) .and. .not. dm_string_has(app%output) .and. .not. app%verbose) then
            call dm_error_out(rc, 'database, output, or verbose mode required')
            return
        end if

        if (.not. dm_log_level_is_valid(app%min_level)) then
            call dm_error_out(rc, 'invalid log level')
            return
        end if

        rc = E_NONE
    end function validate

    ! **************************************************************************
    ! CALLBACKS.
    ! **************************************************************************
    subroutine signal_callback(signum) bind(c)
        !! C-interoperable signal handler that closes database, removes message
        !! queue, and stops program.
        integer(c_int), intent(in), value :: signum

        call logger%debug('exit on on signal ' // dm_posix_signal_name(signum))
        call shutdown(E_NONE)
    end subroutine signal_callback

    subroutine version_callback()
        call dm_version_out(APP_NAME, APP_MAJOR, APP_MINOR, APP_PATCH)
        print '(a, 1x, a)', dm_lua_version(.true.), dm_db_version(.true.)
    end subroutine version_callback
end program dmlogger
